esp_idf_panic_decoder/pc_address_decoder.py

```diff
@@ -23,50 +30,136 @@ def __init__(
+            elf_path = matcher.elf_path
+            is_rom = elf_path == self.rom_elf_file
+            elf_addresses = [addr for addr in remaining if matcher.is_executable_address(int(addr, 16))]
+            if not elf_addresses:
+                continue
+
+            # Lookup addresses using addr2line
+            mapped_addresses = self.lookup_pc_address(elf_addresses, is_rom=is_rom, elf_file=elf_path)
+
+            # Store mapped addresses
+            mapped.update(mapped_addresses)
+
+            # Stop searching for addresses that have been found
+            remaining = [addr for addr in remaining if addr not in mapped_addresses]
+            if not remaining:
+                break
+
+        # Return all mapped addresses that were found, in the original order
+        return [(addr, mapped[addr]) for addr in addresses if addr in mapped]
+
+
+    def lookup_pc_address(
+        self,
+        pc_addr: List[str],
+        is_rom: bool = False,
+        elf_file: str = ''
+    ) -> Dict[str, List[PcAddressLocation]]:
+        """
+        Decode a list of addresses using addr2line, returning a map from each address string
+        to a list of locations (function, path, line).
+        """
+        elf_file = elf_file if elf_file else (self.rom_elf_file if is_rom else self.elf_files[0])  # type: ignore
+        cmd = [f'{self.toolchain_prefix}addr2line', '-fiaC', '-e', elf_file, *pc_addr]
 
         try:
-            translation = subprocess.check_output(cmd, cwd='.')
-            if b'?? ??:0' not in translation:
-                decoded = translation.decode()
-                return decoded if not is_rom else decoded.replace('at ??:?', 'in ROM')
+            batch_output = subprocess.check_output(cmd, cwd='.')
         except OSError as err:
             red_print(f'{" ".join(cmd)}: {err}')
+            return {}
         except subprocess.CalledProcessError as err:
             red_print(f'{" ".join(cmd)}: {err}')
             red_print('ELF file is missing or has changed, the build folder was probably modified.')
-        return None
+            return {}
+
+        decoded_output = batch_output.decode(errors='ignore')
+
+        # Step 1: Split into sections where each section starts with an 8-hex-digit address
+        sections = re.split(ADDR2LINE_ADDRESS_LOOKAHEAD_RE, decoded_output)
+
+        result: Dict[str, List[PcAddressLocation]] = {}
+        for section in sections:
+            section = section.strip()
+            if not section:
+                continue
+
+            # Step 2: Split the section by newlines
+            lines = section.split('\n')
+
+            # Step 3: First line is the address
+            address = lines[0].strip()
+
+            # Step 4: Build trace by consuming lines in pairs (function, path:line)
+            #         Multiple entries indicate inlined functions
+            trace: List[PcAddressLocation] = []
+            for i in range(1, len(lines) - 1, 2):
```

Comment:
If I may suggest this cleaner and more Pythonic alternative:
```python
for fn, path_line in zip(lines[::2], lines[1::2]):
```

And we can embedded stripping there as well:

```python
for fn, path_line in zip(map(str.strip, lines[::2]), map(str.strip, lines[1::2])):
```